README.md
# FastR

FastR is a high-performance implementation of the R programming language, built on GraalVM.

FastR aims to be:
* [efficient](https://medium.com/graalvm/faster-r-with-fastr-4b8db0e0dceb#4ab6): executing R language scripts faster than any other R runtime and as fast as `Rcpp`
* [polyglot](https://medium.com/graalvm/faster-r-with-fastr-4b8db0e0dceb#0f5c): allowing fast [polyglot interoperability](https://www.graalvm.org/22.3/reference-manual/embed-languages/) with other languages in the GraalVM ecosystem.
* [compatible](https://medium.com/graalvm/faster-r-with-fastr-4b8db0e0dceb#fff5): with the reference R implementation including the [R extensions C API](https://cran.r-project.org/doc/manuals/r-release/R-exts.html)
* [embeddable](https://github.com/graalvm/examples/tree/master/r_java_embedding): allowing integration using the R embedding API or the GraalVM polyglot embedding SDK for Java

The screenshot below shows Java application with embedded FastR engine.
The plot below was generated by `ggplot2` running on FastR and it shows peak performance of the [raytracing example](http://www.tylermw.com/throwing-shade/).
The measurements were [reproduced independently](https://web.archive.org/web/20181017111641/https://nextjournal.com/sdanisch/fastr-benchmark).

![Java embedding](documentation/assets/javaui.png)
![Speedup](documentation/assets/speedup.png)

 ## Getting Started

See the documentation on the GraalVM website on how to [get GraalVM](https://www.graalvm.org/22.3/docs/getting-started/) and [install and use FastR](https://www.graalvm.org/22.3/reference-manual/r/).

```
$JAVA_HOME/bin/R
Type 'q()' to quit R.
> print("Hello R!")
[1] "Hello R!"
>
```

## Current Status

The goal of FastR is to be a drop-in replacement for GNU-R, the reference implementation of the R language,
including the [R extensions C API](https://cran.r-project.org/doc/manuals/r-release/R-exts.html).
FastR faithfully implements the R language, and any difference in behavior is considered to be a bug.

### CRAN Packages

FastR can currently install and run basic examples of many of the popular R packages, such as `ggplot2`, `jsonlite`, `testthat`, `assertthat`, `dplyr`, `knitr`, `Shiny`, `Rcpp`, `quantmod`, and more. 
However, one should take into account **the experimental state of FastR**, there can be packages that are not compatible yet, and if you try FastR on a complex R application, it can stumble on those. 
If this happens, please submit an issue on GitHub.

To provide better stability, FastR uses by default a [fixed snapshot](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.native/Makefile#L37) of CRAN (via [MRAN](https://mran.microsoft.com/)). 
Function `install.packages` therefore does not install the latest versions. 
This can be overridden by passing `repos` argument to `install.packages` pointing to CRAN.

FastR provides its own replacements for `rJava` and `data.table` packages, which can be installed with `install.fastr.packages(c("rJava", "data.table"))`.

### Native Extensions Performance

Packages that use the [R extensions C API](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) in hot paths, especially via `Rcpp`, **may exhibit slower performance** on FastR due to the high cost of transitions between the native and managed code.
This can be mitigated by using the [GraalVm LLVM runtime](https://www.graalvm.org/22.3/reference-manual/llvm/). 
Preview of the support is available via the `--R.BackEnd=llvm` option.
Note that most of the times FastR running R code equivalent to given `Rcpp` code is as
fast as GNU-R/Rcpp and sometimes even faster because of the advanced optimizations of the Graal dynamic compiler.

## Documentation

FastR reference documentation which explains its advantages, its current limitations, compatibility, and additional functionality is available on the [GraalVM website](https://www.graalvm.org/22.3/reference-manual/r/).

Further documentation, including contributor and developer-oriented information, is in the [documentation folder](documentation/Index.md) of this repository.

## Stay Connected with the Community

See [graalvm.org/community](https://www.graalvm.org/community/) on how to stay connected with the development community.
The discussion on [Slack](https://www.graalvm.org/slack-invitation/) is a good way to get in touch with us.

We would like to grow the FastR open-source community to provide a free R implementation atop the Truffle/Graal stack.
We encourage contributions, and invite interested developers to join in.
Prospective contributors need to sign the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/).
The access point for contributions, issues and questions about FastR is the [GitHub repository](https://github.com/oracle/fastr).

## Authors

FastR is developed by Oracle Labs and is based on [the GNU-R runtime](http://www.r-project.org/).
It contains contributions by researchers at Purdue University ([purdue-fastr](https://github.com/allr/purdue-fastr)), Northeastern University, JKU Linz, TU Dortmund and TU Berlin.  

## License

FastR is available under a GPLv3 license.


CHANGELOG.md
# 24.0.0
* Migrated the CRAN snapshot, used by default by `install.packages`, to https://packagemanager.posit.co/cran/2021-02-01
  * The MRAN snapshot used previously was shut down
  * The snapshot contents should not change

# 23.0.0
* Removed deprecated FastR grid.
  * Removed `--R.UsetInternalGridGraphics` option.
* Updated `Matrix` recommended package to 1.4-0
* Updated `codetools` recommended package to 0.2-18

# 22.3.0
* Implemented global native variable API, which allows the user to use some native package from two R contexts at the same time.
  * The API consists of a bunch of upcalls, named with a prefix of `FASTR_GlobalVar`, e.g., `FASTR_GlobalVarAlloc`.
  * The documentation is in `com.oracle.truffle.r.ffi.impl.upcalls.FastRUpCalls`.
  * Currently, only `grid` and `graphics` builtin packages are refactored to use the global native variable API.
* Add partial support for `dplyr` 1.0.3
  * Implement `SET_PRCODE`, `SET_PRENV` and `SET_PRVALUE` upcalls.

# 22.2.0
* `--R.DebugLLVMLibs` is no longer a stable option.
* Implemented `SET_GROWABLE_BIT` and `IS_GROWABLE` C API functions.
  * This fixes installation of the `cpp11` 0.2.6 package.
* Add `akima` package to the list of "native packages", so it is by default loaded by the native backend.

# 22.1.0
* Improved performance of the `order` and `rank` builtin functions
* FastR does not ship with its own copy of libz.so
  * System installation of this library becomes a requirement for FastR, but zlib is installed by default on MacOS and in most modern Linux distributions.
* Use JavaGD as the default graphical subsystem.
  * Deprecate `--R.UseInternalGridGraphics` option.
  * The FastR's graphical subsystem is now mostly compatible with GNU-R's, i.e., most functions from `graphics`, `grid`, and `grDevices` base packages are now supported.
  * Display lists are fully implemented.
  * Supported devices: SVG, PNG, JPEG, BMP, AWT.
  * See [graphics docs](./documentation/dev/graphics.md).
* Updated XZ library for compression to the version XZ-1.9.

# 22.0.0
* Adopted [NodeLibrary](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/NodeLibrary.html).
* In unit tests, use single shared context rather than multiple inner contexts.
* Update recommended packages:
  * rpart to version 4.1-16 (2019-05-21)
  * cluster to version 2.1.2 (2021-04-17)

# 21.3.0
* Upgrade of [PCRE](https://www.pcre.org/) to PCRE2 version 10.37.
  * See [GNU-R changelog](https://cran.r-project.org/doc/manuals/r-devel/NEWS.html) (section MIGRATION TO PCRE2) for more details
     on potentially user visible differences between PCRE and PCRE2

Bug fixes:

* Fixed implicit make rule parameters used when building [R extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html)
  * Fixes, e.g., installation of [maps](https://cran.r-project.org/web/packages/maps/index.html) package version 3.3.0.
  * Removed `-shared` flag from `LDFLAGS`, added `-shared` to `DYLIB_LDFLAGS`, added `SHLIB_LDGLAGS_R`
* Fixed unexpected garbage collection of `CHARSXP` objects in [R extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html)
  * FastR did not materialize its internal representation of character vectors to GNU-R compatible CHARSXP objects,
      which caused unexpected collection of CHARSXP objects returned by STRING_ELT
* Option --no-init-file is not ignored anymore (do not read the user's profile at startup)
* Fixed functions `approx` and `approxfun` from the `stats` package
  * Previously they were always failing with error message "Incorrect number of arguments"

# 21.2.0
* Support for packages in 2021-02-01 CRAN snapshot:
  * testthat 3.0.1 is partially supported.
    * FastR does not support parallel tests run, i.e. run testthat only with `Sys.setenv(TESTTHAT_PARALLEL="false")`.
  * tibble 3.0.6 , vctrs 0.3.6, and data.table 1.13.6 are mostly supported.
  * Support for dplyr 1.0.3, ggplot 3.3.3, and knitr 1.31 is a work in progress.
  
Bug fixes:

* `read.dcf` does not ignore whitespaces in fields any more.
* `list.files` gives correct result in a subdirectory with the same prefix as its parent directory.
* Whitespaces in quantifiers in regular expressions are ignored.
  * GNU-R does not comply with PCRE with this behavior.
* `sys.frame` displays frames for `NextMethod` correctly.
* `parent.frame` is able to get the frame that is no longer on the stack.
  * Which is not recommended due to the documentation of `parent.frame`, but some packages do that nonetheless.

# 21.1.0

* Upgraded FastR to R 4.0.3
  * Made FastR runtime mostly compatible with R 4.0.3
  * Migrated to new versions of the base and recommended packages
  * Implemented some of the new features of R 4.0.3
    * All matrices have implicitly also the `array` class
    * Implemented `str2expression`, `str2lang` builtins
    * Pairlists cannot be coerced to language and vice versa
    * Deprecated `--slave` command line option, replaced by `--no-echo`
    * Added `recycle0` parameter to `paste` and `paste0`
    * Add expand parameter to `unlink`
    * `...elt` does no longer accept character vector as parameter
  * See [GNU-R 4.0.3 changelog](https://cran.r-project.org/doc/manuals/r-release/NEWS.pdf) for other changes in the base packages

* Upgraded the CRAN snapshot, used by default by `install.packages`, to 2021-02-01
  * Support of the new versions of most popular packages on FastR is work in progress
  * Packages with known issues: dplyr 1.0.3, ggplot 3.3.3, knitr 1.31

Bug fixes:

* `switch` builtin handles arguments properly (#171)
* `Rf_installTrChar` was giving wrong result
* S3 dispatch in `cbind` and `rbind`
* `.cache_class` with `NULL` parameter
* copy on write of S4 objects when passed into `class<-` function

Added missing R builtins and C APIs:

* `Rf_allocSExp` supports `CLOSXP`
* `R_Serialize` and `R_removeVarFromFrame`


# 21.0.0

Bug fixes:

* Fix `AssertionError` in `sprintf` (#169)

# 20.3.0

Bug fixes:


mx.fastr/suite.py
suite = {
  "mxversion" : "6.0.1",
  "name" : "fastr",
  "versionConflictResolution" : "latest",
  "imports" : {
    "suites" : [
            {
               "name" : "truffle",
               "subdir" : True,
               # The version must be the same as the version of Sulong
               # TRUFFLE REVISION (note: this is a marker for script that can update this)
               "version" : "d24879f0f515f2cc21c318e9cd0ec3f7fd37e71b",
               "urls" : [
                    {"url" : "https://github.com/graalvm/graal", "kind" : "git"},
                    {"url" : "https://curio.ssw.jku.at/nexus/content/repositories/snapshots", "kind" : "binary"},
                ]
            },
            {
               "name" : "sulong",
               "subdir" : True,
               # The version must be the same as the version of Truffle
               # TRUFFLE REVISION (note: this is a marker for script that can update this)
               "version" : "d24879f0f515f2cc21c318e9cd0ec3f7fd37e71b",
               "urls" : [
                    {"url" : "https://github.com/graalvm/graal", "kind" : "git"},
                    {"url" : "https://curio.ssw.jku.at/nexus/content/repositories/snapshots", "kind" : "binary"},
                ]
            },
        ],
   },

  "repositories" : {
    "snapshots" : {
      "url" : "https://curio.ssw.jku.at/nexus/content/repositories/snapshots",
      "licenses" : ["GPLv3"]
    }
  },

  "licenses" : {
    "GPLv3" : {
      "name" : "GNU General Public License, version 3",
      "url" : "https://www.gnu.org/licenses/gpl-3.0.html"
    },
  },

  "defaultLicense" : "GPLv3",

  # libraries that we depend on
  "libraries" : {
    "GNUR" : {
      "path" : "libdownloads/R-4.0.3.tar.gz", # keep in sync with the GraalVM support distribution
      "urls" : ["https://cran.rstudio.com/src/base/R-4/R-4.0.3.tar.gz"],
      "sha1" : "5daba2d63e07a9f39d9b69b68f0642d71213ec5c",
      "resource" : "true"
    },

    "F2C" : {
      "path" : "libdownloads/f2c/src.tgz",
      "urls" : ["https://lafo.ssw.uni-linz.ac.at/pub/graal-external-deps/f2c/20191129/src.tgz"],
      "sha1" : "8a26107bf9f82a2dcfa597f15549a412be75e0ee",
      "resource" : "true"
    },

    "LIBF2C" : {
      "path" : "libdownloads/f2c/libf2c.zip",
      "urls" : ["https://lafo.ssw.uni-linz.ac.at/pub/graal-external-deps/f2c/20191129/libf2c.zip"],
      "sha1" : "e39a00f425f8fc41dde434686080a94e94884f30",
      "resource" : "true"
    },

    # A recommended package "rpart" with a fixed version rather than taken from GNU-R.
    # rpart 4.1-15
    "RPART" : {
      "path" : "libdownloads/rpart.tar.gz",
      "ext" : ".tar.gz",
      "version" : "4020bb4ee8fd6739bd97e8c39931fa7e3901300c",
      "urls" : ["https://api.github.com/repos/bethatkinson/rpart/tarball/{version}"],
      "sha1" : "ec76dbd51acad10bed843a0005ba5fdcf5c7a35d",
      "resource" : "true"
    },

    # A recommended package "cluster" with a fixed version rather than taken from GNU-R.
    "CLUSTER" : {
      "path" : "libdownloads/cluster.tar.gz",
      "ext" : ".tar.gz",
      "version" : "2.1.2",
      "urls" : ["https://cran.r-project.org/src/contrib/Archive/cluster/cluster_{version}.tar.gz"],
      "digest" : "sha512:54f4e4768106035ba0566254393bc392af66cdf9309d117a57d18ee29b93207740d38eac2a4c4db761db65a7300abdbcd2a5915a115d5b81e7f523dd7d583301",
      "resource" : "true"
    },

    # A recommended package "codetools" with a fixed version rather than taken from GNU-R.
    "CODETOOLS" : {
      "path" : "libdownloads/codetools.tar.gz",
      "ext" : ".tar.gz",
      "version" : "0.2-18",
      "urls" : ["https://cran.r-project.org/src/contrib/Archive/codetools/codetools_{version}.tar.gz"],
      "sha1" : "fa0fe4d67316ff49776e5bef1ba56c9334633e71",
      "resource" : "true"
    },

    "MATRIX" : {
      "path" : "libdownloads/matrix.tar.gz",
      "ext" : ".tar.gz",
      "version" : "1.4-0",
      "urls" : ["https://cran.r-project.org/src/contrib/Archive/Matrix/Matrix_{version}.tar.gz"],
      "sha1" : "2745b86754e1becfae6cbea5e4715f87d3fe8464",
      "resource" : "true"
    },

    "BATIK-ALL-1.14" : {
      "digest" : "sha512:6e91f6696e25146109e87ae762b5c0c5c1416680b5c4528bc49a4051497056171b7643d27cbf2e59f2da70fff869409390e17e41dc8adac7c1fd0208aa3115fc",
      "maven" : {
        "groupId" : "org.apache.xmlgraphics",
        "artifactId" : "batik-all",
        "version" : "1.14",
      },
    },
  },

  "projects" : {

    "com.oracle.truffle.r.common" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "sdk:GRAAL_SDK",
      ],
      "requires" : [
        "java.management",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
    },

    "com.oracle.truffle.r.parser" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.runtime",
            "truffle:ANTLR4",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "spotbugsIgnoresGenerated" : True,
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
      "requires" : [
        "java.logging",
      ],
    },

    "com.oracle.truffle.r.nodes" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.runtime",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "annotationProcessors" : [
          "truffle:TRUFFLE_DSL_PROCESSOR",
      ],
      "spotbugsIgnoresGenerated" : True,
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
      "requires" : [
        "java.logging",
        "jdk.unsupported", # sun.misc.Unsafe
      ],
      "forceJavac": "true",
    },

    "com.oracle.truffle.r.nodes.builtin" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.library",
        "sulong:SULONG_API",
      ],
      "requires" : [
        "java.desktop",
        "java.logging",
        "jdk.unsupported", # sun.misc.Unsafe
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "annotationProcessors" : [
        "truffle:TRUFFLE_DSL_PROCESSOR",
      ],
      "spotbugsIgnoresGenerated" : True,
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
      "forceJavac": "true",
    },

    "com.oracle.truffle.r.nodes.test" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.test",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "Truffle,FastR,Test",
      "jacoco" : "include",
      "spotbugsIgnoresGenerated" : True,
    },

    "com.oracle.truffle.r.test" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "mx:JUNIT",
        "truffle:TRUFFLE_TCK",
        "com.oracle.truffle.r.launcher",
        "com.oracle.truffle.r.engine",
      ],
      "requires" : [
        "java.logging",
      ],
      "annotationProcessors" : [
          "truffle:TRUFFLE_DSL_PROCESSOR",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "Truffle,FastR,Test",
      "jacoco" : "include",
      "spotbugsIgnoresGenerated" : True,
    },

    "com.oracle.truffle.r.test.native" : {
      "native" : True,
      "sourceDirs" : [],
      "dependencies" : ["com.oracle.truffle.r.native"],
      "platformDependent" : True,
      "output" : "com.oracle.truffle.r.test.native",
      "buildEnv" : {
        "LABS_LLVM_CC": "<toolchainGetToolPath:native,CC>",
        "LABS_LLVM_CXX": "<toolchainGetToolPath:native,CXX>",
      },
      "results" :[
         "urand/lib/liburand.so",
       ],
      "workingSets" : "FastR",
      "spotbugsIgnoresGenerated" : True,
    },

    "com.oracle.truffle.r.test.packages" : {
      "sourceDirs" : ["r"],
      "javaCompliance" : "17+",
      "workingSets" : "FastR",
    },

    "com.oracle.truffle.r.test.packages.analyzer" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "mx:JUNIT"
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "FastR",
      "requires" : [
        "java.logging"
      ],
    },

    "com.oracle.truffle.r.engine" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.nodes.builtin",
        "com.oracle.truffle.r.parser",
        "sdk:JLINE3",
        "truffle:TRUFFLE_NFI",
      ],
     "generatedDependencies" : [
        "com.oracle.truffle.r.parser",
     ],
      "annotationProcessors" : [
          "truffle:TRUFFLE_DSL_PROCESSOR",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
      "spotbugsIgnoresGenerated" : True,
      "requires" : [
        "java.logging",
      ],
    },

    "com.oracle.truffle.r.runtime" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.common",
        "truffle:TRUFFLE_API",
        "sulong:SULONG_API",
        "truffle:TRUFFLE_XZ",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "checkstyleVersion": "8.8",
      "javaCompliance" : "17+",
      "annotationProcessors" : [
          "truffle:TRUFFLE_DSL_PROCESSOR",
      ],
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
      "spotbugsIgnoresGenerated" : True,
      "requires" : [
        "java.management",
        "java.desktop",
        "java.logging",
        "jdk.unsupported" # sun.misc.Unsafe
      ],
    },

    "com.oracle.truffle.r.launcher" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.common",
        "sdk:GRAAL_SDK",
        "sdk:LAUNCHER_COMMON",
        "sdk:JLINE3",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
    },

    "com.oracle.truffle.r.ffi.impl" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
         "com.oracle.truffle.r.ffi.processor",
         "com.oracle.truffle.r.nodes",
         "org.rosuda.javaGD",
         "truffle:TRUFFLE_NFI",
         'BATIK-ALL-1.14',
      ],
      "requires" : [
        "java.desktop",
        "jdk.unsupported" # sun.misc.Unsafe
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "annotationProcessors" : [
          "truffle:TRUFFLE_DSL_PROCESSOR",
          "R_FFI_PROCESSOR",
      ],
      "workingSets" : "Truffle,FastR",
      "jacoco" : "include",
      "spotbugsIgnoresGenerated" : True,
    },

    "com.oracle.truffle.r.ffi.codegen" : {
      "sourceDirs" : ["src"],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "dependencies" : [
        "com.oracle.truffle.r.ffi.impl"
      ],
      "javaCompliance" : "17+",
      "workingSets" : "FastR",
    },

    "com.oracle.truffle.r.ffi.processor" : {
      "sourceDirs" : ["src"],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "FastR",
      "requires" : [
        "java.compiler"
      ],
    },

    "com.oracle.truffle.r.native" : {
      "sourceDirs" : [],
      "dependencies" : [
        "GNUR",
        "F2C",
        "LIBF2C",
        "truffle:TRUFFLE_NFI_NATIVE",
        "sulong:SULONG_BOOTSTRAP_TOOLCHAIN",
        "sulong:SULONG_HOME",
        "sulong:SULONG_LEGACY",
      ],
      "native" : True,
      "single_job" : True,
      "workingSets" : "FastR",
      "buildEnv" : {
        "NFI_INCLUDES" : "-I<path:truffle:TRUFFLE_NFI_NATIVE>/include",
        "LLVM_INCLUDES" : "-I<path:sulong:SULONG_LEGACY>/include -I<path:sulong:SULONG_HOME>/include",
        "LLVM_LIBS_DIR" : "<path:sulong:SULONG_HOME>",
        # If FASTR_RFFI=='llvm', then this is set as CC/CXX in c.o.t.r.native/Makefile
        "LABS_LLVM_CC": "<toolchainGetToolPath:native,CC>",
        "LABS_LLVM_CXX": "<toolchainGetToolPath:native,CXX>",
        "GRAALVM_VERSION": "<graalvm_version>",
      },
    },

    "com.oracle.truffle.r.library" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "com.oracle.truffle.r.ffi.impl",
      ],
      "requires" : [
        "java.desktop",
        "jdk.unsupported", # sun.misc.Unsafe
      ],
      "annotationProcessors" : [
          "truffle:TRUFFLE_DSL_PROCESSOR",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "FastR",
      "jacoco" : "include",
      "spotbugsIgnoresGenerated" : True,
    },

    "org.rosuda.javaGD" : {
      "sourceDirs" : ["src"],
      "dependencies" : [],
      "requires": ["java.desktop"],
      "checkstyle" : "org.rosuda.javaGD",
      "javaCompliance" : "17+",
      "workingSets" : "FastR",
      "jacoco" : "include",
    },

    "com.oracle.truffle.r.release" : {
      "sourceDirs" : ["src"],
      "buildDependencies" : ["com.oracle.truffle.r.native.recommended"],
      "class" : "FastRReleaseProject",
      "output" : "com.oracle.truffle.r.release",
    },

    "com.oracle.truffle.r.native.recommended" : {
      "dependencies" : [
        "RPART",
        "CLUSTER",
        "CODETOOLS",
        "MATRIX",
        "com.oracle.truffle.r.native",
        "com.oracle.truffle.r.engine",
        "com.oracle.truffle.r.ffi.impl",
        "com.oracle.truffle.r.common",
      ],
      "max_jobs" : "8",
      "native" : True,
      "vpath": True,
      "workingSets" : "FastR",
      "buildDependencies" : ["FASTR"],
    },

    "com.oracle.truffle.r.test.tck" : {
      "sourceDirs" : ["src"],
      "dependencies" : [
        "mx:JUNIT",
        "sdk:POLYGLOT_TCK",
      ],
      "checkstyle" : "com.oracle.truffle.r.runtime",
      "javaCompliance" : "17+",
      "workingSets" : "FastR,Test",
      "spotbugsIgnoresGenerated" : True,
    },
  },

  "distributions" : {
    "R_FFI_PROCESSOR" : {
      "description" : "internal support for generating FFI classes",
      "dependencies" : ["com.oracle.truffle.r.ffi.processor"],
      "maven" : "False",
      # FASTR and R_FFI_PROCESSOR share the actual annotations
      # This could be refactored so that we have one project with just the annotations and FASTR would depend only on that
      "overlaps": ["FASTR"],
    },

    "FASTR_COMMON" : {
      "moduleInfo" : {
          "name" : "org.graalvm.r.common",
          "exports" : [
              "com.oracle.truffle.r.common to org.graalvm.r",
            "com.oracle.truffle.r.common to org.graalvm.r.launcher",
          ],
      },
      "useModulePath": True,
      "description" : "common utilities shared by fastr and fastr-launcher",
      "dependencies" : ["com.oracle.truffle.r.common"],
      "distDependencies" : [
        "sdk:GRAAL_SDK",
      ],
    },

    "FASTR_LAUNCHER" : {
      "moduleInfo" : {
        "name" : "org.graalvm.r.launcher",
      },
      "useModulePath": True,
      "description" : "launcher for the GraalVM (at the moment used only when native image is installed)",
      "dependencies" : ["com.oracle.truffle.r.launcher"],
      "distDependencies" : [
        "FASTR_COMMON",
        "sdk:GRAAL_SDK",
        "sdk:LAUNCHER_COMMON",
        "sdk:JLINE3",
      ],
    },

    "FASTR" : {
      "moduleInfo" : {
          "name" : "org.graalvm.r",
      },
      "useModulePath": True,
      "requires": [
          "java.base",
          "java.logging",
          "java.management",
          "jdk.unsupported",
      ],
      "description" : "class files for compiling against FastR in a separate suite",
      "dependencies" : [
        "com.oracle.truffle.r.engine",
        "com.oracle.truffle.r.ffi.impl"
      ],
      "mainClass" : "com.oracle.truffle.r.launcher.RCommand",
      "exclude" : [
        "truffle:ANTLR4",
        "GNUR",
        "BATIK-ALL-1.14",
      ],
      "distDependencies" : [
        "FASTR_COMMON",
        "truffle:TRUFFLE_API",
        "truffle:TRUFFLE_NFI",
        "truffle:TRUFFLE_NFI_LIBFFI",
        "truffle:TRUFFLE_XZ",
        "sulong:SULONG_API",
        "sdk:JLINE3",
      ],
    },

    "FASTR_COMMUNITY" : {
      "type" : "pom",
      "runtimeDependencies" : [
        "fastr:FASTR_COMMON",
        "fastr:FASTR",
      ],
      "description" : "FastR support distribution for the GraalVM",
      "maven" : False,
    },

    "FASTR_UNIT_TESTS" : {
      "description" : "unit tests",
      "dependencies" : [
        "com.oracle.truffle.r.test",
        "com.oracle.truffle.r.nodes.test"
       ],
      "exclude": ["mx:HAMCREST", "mx:JUNIT"],
      "distDependencies" : [
        "FASTR",
        "FASTR_LAUNCHER",
        "sulong:SULONG_NATIVE",
        "truffle:TRUFFLE_API",
        "truffle:TRUFFLE_TCK",
      ],
      "unittestConfig": "fastr-tests",
    },

    "FASTR_UNIT_TESTS_NATIVE" : {
      "description" : "unit tests support (from test.native project)",
       "native" : True,
       "platformDependent" : True,
      "dependencies" : [
        "com.oracle.truffle.r.test.native",
     ],
    },

    "TRUFFLE_R_TCK" : {
      "description" : "TCK tests provider",
      "dependencies" : [
        "com.oracle.truffle.r.test.tck"
      ],
      "exclude" : [
        "mx:JUNIT",
      ],
      "distDependencies" : [
        "sdk:POLYGLOT_TCK",
        "FASTR",
      ],
      "maven" : False
    },

    # see mx_fastr_dists.mx_register_dynamic_suite_constituents for the definitions of some RFFI-dependent distributions
  },
}
